README.md
# nsd cookbook
[![Chef cookbook](https://img.shields.io/cookbook/v/nsd.svg?style=flat-square)]()
[![license](https://img.shields.io/github/license/aspyatkin/nsd-cookbook.svg?style=flat-square)]()  
A Chef cookbook to install nsd master and slaves along with initial zone configuration.

## Resources

The examples below configure one master (`ns1.example.com`) and two slave servers (`ns2.example.com` and `ns3.example.com`), all of which host two zones - `example.com` itself and `test.com`. SOA records will contain `hostmaster@example.com` as an email address.

### nsd_master

```ruby
nsd_master 'ns1.example.com' do
  port 53
  ipv4_address '1.1.1.1'
  ipv6_address '2001:0db8:0a0b:12f0:0000:0000:0000:0001'
  contact 'hostmaster@example.com'
  bind_addresses %w(1.1.1.1 2001:0db8:0a0b:12f0:0000:0000:0000:0001)
  enable_ipv6 true
  keys({
    'ns2.example.com' => '7DzUnLpx9H...',
    'ns3.example.com' => '1ADEn1fqOo...'
  })
  slaves({
    'ns2.example.com' => {
      'ipv4_address' => '2.2.2.2',
      'ipv6_address' => '2001:0db8:0a0b:12f0:0000:0000:0000:0002'
    },
    'ns3.example.com' => {
      'ipv4_address' => '3.3.3.3',
      'ipv6_address' => '2001:0db8:0a0b:12f0:0000:0000:0000:0003'
    }
  })
  zones %w(example.com test.com)
  action :install
end
```

### nsd_slave

```ruby
nsd_slave 'ns2.example.com' do
  port 53
  bind_addresses %w(2.2.2.2 2001:0db8:0a0b:12f0:0000:0000:0000:0002)
  enable_ipv6 true
  key '7DzUnLpx9H...'
  zones %w(example.com test.com)
  master_ipv4_address '1.1.1.1'
  action :install
end

```

```ruby
nsd_slave 'ns3.example.com' do
  port 53
  bind_addresses %w(3.3.3.3 2001:0db8:0a0b:12f0:0000:0000:0000:0003)
  enable_ipv6 true
  key '1ADEn1fqOo...'
  zones %w(example.com test.com)
  master_ipv4_address '1.1.1.1'
  action :install
end

```

## AXFR keys

A zone is modified only on a master server. Then the changes are transferred to slave servers. It is presupposed that each slave server has its own secret key. Secret management is up to the user utilizing `nsd_master` and `nsd_slave` resources.

A key may be generated with this command:

```sh
$ dd if=/dev/random count=1 bs=32 2> /dev/null | base64
```

## Limitations
This cookbook does not set up any firewall rules.

## License
MIT @ [Alexander Pyatkin](https://github.com/aspyatkin)


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [0.5.0] - 2021-03-22

### Added
- Add CHANGELOG.

### Changed
- Change zone file structure.


resources/zone.rb
require 'date'

resource_name :nsd_zone

property :fqdn, String, name_property: true
property :zone_dir, String, required: true
property :master_fqdn, String, required: true
property :master_ipv4_address, String, required: true
property :master_ipv6_address, [String, NilClass], default: nil
property :master_contact, String, required: true
property :slaves, Hash, required: true
property :refresh, Integer, default: 3_600
property :retry, Integer, default: 900
property :expire, Integer, default: 1_209_600
property :ttl, Integer, default: 600

default_action :create

action :create do
  zone_file = ::File.join(new_resource.zone_dir, new_resource.fqdn, "#{new_resource.fqdn}.zone")

  directory ::File.dirname(zone_file) do
    owner 'root'
    group node['root_group']
    mode 0755
    action :create
  end

  template zone_file do
    cookbook 'nsd'
    source 'nsd.zone.erb'
    owner 'root'
    group node['root_user']
    variables(
      fqdn: new_resource.fqdn,
      master_fqdn: new_resource.master_fqdn,
      master_ipv4_address: new_resource.master_ipv4_address,
      master_ipv6_address: new_resource.master_ipv6_address,
      master_contact: new_resource.master_contact,
      slaves: new_resource.slaves,
      serial: ::DateTime.now.strftime('%Y%m%d01'),
      refresh: new_resource.refresh,
      retry: new_resource.retry,
      expire: new_resource.expire,
      ttl: new_resource.ttl
    )
    mode 0644
    action :create_if_missing
  end
end


resources/slave.rb
resource_name :nsd_slave
property :fqdn, String, name_property: true

property :user, String, default: 'nsd'
property :group, String, default: 'nsd'
property :uid, Integer, default: 500
property :gid, Integer, default: 500

property :port, Integer, default: 53
property :service, String, default: 'nsd'
property :conf_dir, String, default: '/etc/nsd'
property :log_file, String, default: '/var/log/nsd'
property :pid_file, String, default: '/run/nsd/nsd.pid'

property :bind_addresses, Array, required: true
property :enable_ipv6, [TrueClass, FalseClass], default: false
property :key, String, required: true
property :zones, Array, required: true
property :master_ipv4_address, String, required: true

default_action :install

action :install do
  nsd_user new_resource.user do
    group new_resource.group
    uid new_resource.uid
    gid new_resource.gid
    action :create
  end

  systemd_unit new_resource.service do
    action :nothing
  end
  systemd_unit_resource = "systemd_unit[#{new_resource.service}]"

  service new_resource.service do
    action :nothing
  end
  service_resource = "service[#{new_resource.service}]"

  %w(
    nsd
  ).each do |pkg_name|
    package pkg_name do
      action :install
      notifies :mask, systemd_unit_resource, :before
      notifies :unmask, systemd_unit_resource, :immediately
    end
  end

  zone_dir = ::File.join(new_resource.conf_dir, 'zones')

  directory zone_dir do
    owner 'root'
    group node['root_group']
    mode 0755
    action :create
  end

  template ::File.join(new_resource.conf_dir, 'nsd.conf') do
    cookbook 'nsd'
    source 'nsd.slave.conf.erb'
    owner 'root'
    group node['root_group']
    variables(
      port: new_resource.port,
      bind_addresses: new_resource.bind_addresses,
      enable_ipv6: new_resource.enable_ipv6,
      service_user: new_resource.user,
      zone_dir: zone_dir,
      log_file: new_resource.log_file,
      pid_file: new_resource.pid_file,
      key_name: new_resource.fqdn,
      key_secret: new_resource.key,
      zones: new_resource.zones,
      master_ipv4_address: new_resource.master_ipv4_address
    )
    mode 0644
    sensitive true
    action :create
    notifies :restart, "service[#{new_resource.service}]", :delayed
  end

  service new_resource.service do
    action [:start, :enable]
  end
end


resources/master.rb
resource_name :nsd_master
property :fqdn, String, name_property: true

property :user, String, default: 'nsd'
property :group, String, default: 'nsd'
property :uid, Integer, default: 500
property :gid, Integer, default: 500

property :port, Integer, default: 53
property :service, String, default: 'nsd'
property :conf_dir, String, default: '/etc/nsd'
property :log_file, String, default: '/var/log/nsd'
property :pid_file, String, default: '/run/nsd/nsd.pid'

property :ipv4_address, String, required: true
property :ipv6_address, [String, NilClass], default: nil
property :contact, String, required: true
property :bind_addresses, Array, required: true
property :enable_ipv6, [TrueClass, FalseClass], default: false
property :keys, Hash, required: true
property :slaves, Hash, required: true
property :zones, Array, required: true

default_action :install

action :install do
  nsd_user new_resource.user do
    group new_resource.group
    uid new_resource.uid
    gid new_resource.gid
    action :create
  end

  systemd_unit new_resource.service do
    action :nothing
  end
  systemd_unit_resource = "systemd_unit[#{new_resource.service}]"

  service new_resource.service do
    action :nothing
  end
  service_resource = "service[#{new_resource.service}]"

  %w(
    nsd
  ).each do |pkg_name|
    package pkg_name do
      action :install
      notifies :mask, systemd_unit_resource, :before
      notifies :unmask, systemd_unit_resource, :immediately
    end
  end

  zone_dir = ::File.join(new_resource.conf_dir, 'zones')

  directory zone_dir do
    owner 'root'
    group node['root_group']
    mode 0755
    action :create
  end

  template ::File.join(new_resource.conf_dir, 'nsd.conf') do
    cookbook 'nsd'
    source 'nsd.master.conf.erb'
    owner 'root'
    group node['root_group']
    variables(
      port: new_resource.port,
      bind_addresses: new_resource.bind_addresses,
      enable_ipv6: new_resource.enable_ipv6,
      service_user: new_resource.user,
      zone_dir: zone_dir,
      log_file: new_resource.log_file,
      pid_file: new_resource.pid_file,
      keys: new_resource.keys,
      slaves: new_resource.slaves,
      zones: new_resource.zones
    )
    mode 0644
    sensitive true
    action :create
    notifies :restart, "service[#{new_resource.service}]", :delayed
  end

  new_resource.zones.each do |zone_fqdn|
    nsd_zone zone_fqdn do
      zone_dir zone_dir
      master_fqdn new_resource.fqdn
      master_ipv4_address new_resource.ipv4_address
      master_ipv6_address new_resource.ipv6_address
      master_contact new_resource.contact
      slaves new_resource.slaves
      action :create
      notifies :restart, "service[#{new_resource.service}]", :delayed
    end
  end

  service new_resource.service do
    action [:start, :enable]
  end
end
